### ISSUE TITLE
Android app fails due to missing dependency

### ISSUE THREAD

**Issue by AArnott** [AArnott/Avalonia-WithNativeDependency#1]
On the line that first requires the native dependency to be loaded, an exception is thrown: 

```
System.DllNotFoundException
  Message=nerdbank_zcash_rust
 	0xFFFFFFFFFFFFFFFF in Android.Runtime.RuntimeNativeMethods.monodroid_debugger_unhandled_exception	C#
 	0x1A in Android.Runtime.JNINativeWrapper._unhandled_exception at /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:13,5	C#
 	0x1E in Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PPL_V at /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:126,26	C#
 	0x1A in NativeMethods.get_orchard_fvk_bytes_from_sk_bytes	C#
 	0x16 in NativeMethods.TryDeriveOrchardFullViewingKeyFromSpendingKey	C#
 	0x14 in Nerdbank.Zcash.Orchard.SpendingKey.CreateFullViewingKey	C#
 	0x1B in Nerdbank.Zcash.Orchard.SpendingKey..ctor	C#
 	0x55 in Nerdbank.Zcash.Zip32HDWallet.Orchard.Create	C#
 	0x34 in Nerdbank.Zcash.Zip32HDWallet..ctor	C#
 	0x17 in Nerdbank.Zcash.Zip32HDWallet..ctor	C#
>	0x13 in AvaloniaApplication1.ViewModels.MainViewModel..ctor at c:\temp\AvaloniaApplication1\AvaloniaApplication1\ViewModels\MainViewModel.cs:12,3	C#
 	0x4D in AvaloniaApplication1.App.OnFrameworkInitializationCompleted at c:\temp\AvaloniaApplication1\AvaloniaApplication1\App.axaml.cs:28,13	C#
 	0xC1 in Avalonia.AppBuilder.SetupUnsafe	C#
 	0x65 in Avalonia.AppBuilder.Setup	C#
 	0x8 in Avalonia.AppBuilder.SetupWithLifetime	C#
 	0x13 in Avalonia.Android.AvaloniaMainActivity<AvaloniaApplication1.App>.InitializeApp	C#
 	0x1 in Avalonia.Android.AvaloniaMainActivity<AvaloniaApplication1.App>.OnCreate	C#
 	0x11 in Android.App.Activity.n_OnCreate_Landroid_os_Bundle_ at /Users/runner/work/1/s/xamarin-android/src/Mono.Android/obj/Release/net8.0/android-34/mcw/Android.App.Activity.cs:3082,4	C#
 	0x9 in Android.Runtime.JNINativeWrapper.Wrap_JniMarshal_PPL_V at /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Android.Runtime/JNINativeWrapper.g.cs:125,5	C#
```

This, despite the fact that the dll _does_ in fact ship with the app alongside the other binaries:
![image](https://github.com/AArnott/Avalonia-WithNativeDependency/assets/3548/d381e50c-9ef7-419a-be2e-d4b7de1f5a65)

Further, I've confirmed with the msbuild binlog file that the dll that's placed there is the correct android-x64 build of the native binary, which given this is running in an x64 emulator seems to be the right one to 'ship' with the app for the emulator.

**Comment by AArnott** [comment]
I'm offering a $150 bounty (payable in Zcash, Litecoin, Venmo or PayPal) to the first party to submit a PR (or adequately prescriptive comment here) that fixes the problem.

**Comment by maxkatz6** [comment]
This was an instruction for old Xamarin.Native for Android: https://learn.microsoft.com/en-us/xamarin/android/platform/native-libraries

Microsoft never really documented the same for .NET (Core) mobile workloads, but the same approach should still work. 
Alternatively, they have is kinda documented on GitHub page: https://github.com/xamarin/xamarin-android/blob/aea89e161fb49fe921d9c601abb44360c653787d/Documentation/guides/OneDotNetEmbeddedResources.md

In summary, you likely only need `AndroidNativeLibrary` ItemGroup with correct Abi.

**Comment by AArnott** [comment]
Thanks. I'd already found and read the doc on native-libraries on learn.microsoft.com. So far, it hasn't helped. When I created an `AndroidNativeLibrary` item, msbuild told me it was a duplicate and would be ignored. This makes sense to me, because the NuGet PackageReference I have should have been turned into the necessary item automatically.

>  warning XA4301: APK already contains th
e item lib/x86_64/libnerdbank_zcash_rust.so; ignoring.

Having left out the `Abi` metadata, the build targets are left to sniff the path for hints. It isn't complaining about this, but all it has to go on is `android-x64`. I can't tell from the binlog what ABI this equates to. Rust doesn't document the ABI it produces given a target of `x86_64-linux-android` either, or I'd try adding the msbuild item explicitly to see if specifying the Abi metadata would help.

**Comment by AArnott** [comment]
Adding this didn't solve the problem, and the warning reappeared.

```xml
<AndroidNativeLibrary Include="..\..\..\Users\andre\.nuget\packages\nerdbank.zcash\0.2.109-beta\runtimes\android-x64\native\libnerdbank_zcash_rust.so" Link="libnerdbank_zcash_rust.so">
  <Abi>x86_64</Abi>
</AndroidNativeLibrary>
```

**Comment by AArnott** [comment]
I could suppress the warning by adding `ExcludeAssets="native"` to my `PackageReference`. But it didn't solve the runtime failure.

**Comment by maxkatz6** [comment]
Yesterday I didn't notice that this native library was part of the nuget already.

Today I tried couple of things:
1. Checked logcat logs - nothing useful. Only stuff like `Shared library 'libnerdbank_zcash_rust' not loaded, p/invoke 'get_orchard_fvk_bytes_from_sk_bytes' may fail`.
2. Tried to extract `get_orchard_fvk_bytes_from_sk_bytes` Dllimport out of the nuget to simplify testing. With or without "lib" prefix - no difference.
3. Tried to peek with NativeLibrary.Load - no difference.
4. Tried to manually include this ".so" file similarly on how SkiaSharp [does it](https://github.com/mono/SkiaSharp/blob/7ebc1f0719e68210523355fadd8dec4a9ef5f34c/binding/IncludeNativeAssets.SkiaSharp.targets#L30):
```
<PackageReference Include="Nerdbank.Zcash" Version="0.2.109-beta" GeneratePathProperty="true" />
<AndroidNativeLibrary Include="$(PkgNerdbank_Zcash)\runtimes\android-x64\native\libnerdbank_zcash_rust.so" Abi="x86_64" />
<AndroidNativeLibrary Include="$(PkgNerdbank_Zcash)\runtimes\android-arm64\native\libnerdbank_zcash_rust.so" Abi="arm64-v8a" />
```
5. Explicitly specifying `<RuntimeIdentifier>android-x64</RuntimeIdentifier>` also does nothing.
6. Checked the apk file - libnerdbank_zcash_rust.so is there as expected, in x86_64 folder.
7. With `nm -D libnerdbank_zcash_rust.so` checked if library has needed exports - it does.

**Comment by maxkatz6** [comment]
Might be worth to ask [xamarin-android](https://github.com/xamarin/xamarin-android) guys, if they know what's going on here.
Possibly there is some problem or incompatibility with how Rust builds the .so file, but I don't know much about that.

**Comment by AArnott** [comment]
Thanks for trying all that. FWIW, I tried libgit2sharp too, which has a native library too (which isn't rust AFAIK). It failed in the same way.
Thanks for the link to the xamarin-android folks. I'll try there.

**Comment by AArnott** [comment]
It turns out to work fine on a real Android device. It's just the emulator that fails. Go figure.

**Comment by hxp0571-creator** [comment]
I think this is now diagnosable/solvable without changing the Avalonia sample itself.

Root cause: the Android native library was being selected/packaged, but the x86_64 emulator failed to `dlopen` `libnerdbank_zcash_rust.so` because the Rust-built android-x64 binary referenced the compiler runtime helper symbol `__extenddftf2`. The managed exception then surfaced as `DllNotFoundException`, which made it look like an APK packaging problem. The physical arm64 device working is consistent with this being specific to the android-x64 native artifact.

Evidence chain:

- The upstream `dotnet/android` follow-up found the real loader error: `dlopen failed: cannot locate symbol "__extenddftf2" referenced by .../lib/x86_64/libnerdbank_zcash_rust.so` in https://github.com/dotnet/android/issues/8788#issuecomment-1980765624.
- `llvm-nm -D` is the right verification loop: before the Rust workaround, the android-x64 `.so` contained `U __extenddftf2`; after the workaround, that undefined symbol disappeared, per https://github.com/dotnet/android/issues/8788#issuecomment-1996330258.
- The workaround was merged in `Nerdbank.Cryptocurrencies` here: https://github.com/nerdcash/Nerdbank.Cryptocurrencies/pull/262. In short, the Android build script needs to link the NDK Clang builtins runtime for the Android target, especially android-x64.

Prescriptive fix/checklist:

1. Build the Nerdbank.Zcash Rust native library with the PR 262-style Android linker setup so android-x64 links `clang_rt.builtins-x86_64-android` from the NDK.
2. Verify the produced `runtimes/android-x64/native/libnerdbank_zcash_rust.so` with:

   ```sh
   $ANDROID_NDK_HOME/toolchains/llvm/prebuilt/<host>/bin/llvm-nm -D runtimes/android-x64/native/libnerdbank_zcash_rust.so | grep __extenddftf2
   ```

   Expected result: no `U __extenddftf2`.
3. Make sure the app is not deploying a stale native binary: uninstall the app from the emulator, delete `bin/` and `obj/`, rebuild with `dotnet build -bl`, install the resulting APK with `adb install`, and inspect the binlog if needed to confirm the android-x64 `.so` came from the repaired package.

I also statically checked the current public `Nerdbank.Zcash` 0.4.9 package locally: its android-x64 `.so` no longer contains the `__extenddftf2` string, and both Android native libraries only list `libdl.so`, `libm.so`, and `libc.so` as `NEEDED` dependencies. That is consistent with the upstream diagnosis and the merged workaround.

If this counts as the adequately prescriptive bounty answer, PayPal is https://PayPal.Me/hxp8848.
